widget/embedded/src/utils/colors.ts

```diff
@@ -1,141 +1,185 @@
+  return calculateColors(hex, generateShade);
 }
 
-// Function to clamp color component to valid range [0, 255] and convert to hex
-function clamp(color: number): string {
-  const MAX_COLOR_VALUE = 255;
-  const HEX_BASE = 16;
-  const HEX_LENGTH = -2;
-  return (
-    '0' + Math.max(0, Math.min(MAX_COLOR_VALUE, color)).toString(HEX_BASE)
-  )
-    .slice(HEX_LENGTH)
-    .toUpperCase();
+// Calculate an array of tints from a hex color string
+export function calculateTints(hex: string): string[] {
+  return calculateColors(hex, generateTint);
 }
 
-export function isBright(baseColor: string) {
-  const BRIGHTNESS_THRESHOLD = 188;
-  const RED_COEFFICIENT = 299;
-  const GREEN_COEFFICIENT = 587;
-  const BLUE_COEFFICIENT = 114;
-  const BRIGHTNESS_DIVISOR = 1000;
-
-  const [r, g, b] = convertToRGB(baseColor);
-  const brightness =
-    (r * RED_COEFFICIENT + g * GREEN_COEFFICIENT + b * BLUE_COEFFICIENT) /
-    BRIGHTNESS_DIVISOR;
-  if (brightness > BRIGHTNESS_THRESHOLD) {
-    return true;
+// General function to calculate shades or tints
+function calculateColors(
+  hex: string,
+  generateColor: ShadeOrTintFunction
+): string[] {
+  const rgb = hexToRGB(hex);
+  const colors: string[] = [];
+  for (let i = 1; i <= 9; i++) {
+    colors.push(rgbToHex(generateColor(rgb, i)));
   }
-  return false;
+  return colors;
 }
 
-export const generateColors = (
-  mainColors: { [x: string]: string },
-  reverseNeutralRange: boolean,
-  colors?: WidgetColors
-) => {
-  if (!colors || !Object.entries(colors).length) {
-    return {};
+// Create tints and shades for a given color
+export function createTintsAndShades(
+  hex: string,
+  colorKey: string,
+  isReversed?: boolean
+): { [key: string]: string } {
+  const BASE_INDEX = 100;
+  const STEP = 50;
+
+  const tints = calculateTints(hex).reverse().concat(hex);
+  const shades = calculateShades(hex);
+  const combinedColors = tints.concat(shades);
+
+  const colorMap: { [key: string]: string } = {};
+  const length = combinedColors.length;
+
+  for (let i = 0; i < length; i++) {
+    const index = BASE_INDEX + (isReversed ? length - 1 - i : i) * STEP;
+    colorMap[`${colorKey}${index}`] = combinedColors[i];
   }
-  let changeColors = false;
-  let listOfColors = { ...mainColors };
-  for (const colorKey in colors) {
-    const color = colors[colorKey as WidgetColorsKeys];
-
-    if (!!color && color !== mainColors[colorKey]) {
-      changeColors = true;
-      if (colorKey === 'background' || colorKey === 'foreground') {
-        listOfColors = {
-          ...listOfColors,
-          [colorKey]: color,
-        };
-      } else {
-        const reverse = colorKey === 'neutral' && reverseNeutralRange;
-        listOfColors = {
-          ...listOfColors,
-          ...generateShades(color, colorKey, reverse),
-        };
-      }
+
+  return colorMap;
+}
+
+// Generate a color palette based on base colors and "Expand Colors"
+export function expandedThemeColors(
```

Comment:
I think `expandToGenerateThemeColors` is more meaningful. Sorry for keep changing this function name for multiple times, trying to find the best name to describe what it's doing.